client.go

```diff
@@ -25,53 +28,160 @@ const (
 // Client represents the client of an identity server.
 type Client struct {
 	client
+
+	// permChecker is used to check group membership.
+	// It is only non-zero when groups are enabled.
+	permChecker *PermChecker
 }
 
+var _ IdentityClient = (*Client)(nil)
+
 // NewParams holds the parameters for creating a new client.
 type NewParams struct {
+	// BaseURL holds the URL of the identity manager.
 	BaseURL string
-	Client  *httpbakery.Client
 
-	// AuthUsername holds the username for admin login.
-	AuthUsername string
+	// Client holds the client to use to make requests
+	// to the identity manager.
+	Client *httpbakery.Client
 
-	// AuthPassword holds the password for admin login.
-	AuthPassword string
+	// AgentUsername holds the username for group-fetching authorization.
+	// If this is empty, no group information will be provided.
+	// The agent key is expected to be held inside the Client.
+	AgentUsername string
+
+	// CacheTime holds the maximum duration for which
+	// group membership information will be cached.
+	// If this is zero, group membership information will not be cached.
+	CacheTime time.Duration
 }
 
 // New returns a new client.
 func New(p NewParams) *Client {
 	var c Client
 	c.Client.BaseURL = p.BaseURL
-	if p.AuthUsername != "" {
-		c.Client.Doer = &basicAuthClient{
-			client:   p.Client,
-			user:     p.AuthUsername,
-			password: p.AuthPassword,
+	if p.AgentUsername != "" {
+		// Set up agent auth.
+		u, err := url.Parse(p.BaseURL)
+		if err != nil {
+			panic(err)
```

Comment:
panicking seems such an extreme solution, when we can return an error and let the caller deal with the aftermath.